internal/credspropagation/openstack.go

```diff
@@ -0,0 +1,273 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package credspropagation
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	texttemplate "text/template"
+
+	"gopkg.in/yaml.v3"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	capo "sigs.k8s.io/cluster-api-provider-openstack/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+)
+
+// Fields needed to populate the cloud.conf template
+type CloudConfFields struct {
+	AuthURL                     string
+	ApplicationCredentialID     string
+	ApplicationCredentialName   string
+	ApplicationCredentialSecret string
+	Username                    string
+	Password                    string
+	ProjectID                   string
+	ProjectName                 string
+	UserDomainName              string
+	ProjectDomainName           string
+	RegionName                  string
+	Interface                   string
+	FloatingNetworkID           string
+	PublicNetworkName           string
+}
+
+type CloudsYaml struct {
+	Clouds map[string]Cloud `yaml:"clouds"`
+}
+
+type Cloud struct {
+	Auth       Auth   `yaml:"auth"`
+	RegionName string `yaml:"region_name"`
+	Interface  string `yaml:"interface"`
+}
+
+type Auth struct {
+	AuthURL                     string `yaml:"auth_url"`
+	ApplicationCredentialID     string `yaml:"application_credential_id"`
+	ApplicationCredentialName   string `yaml:"application_credential_name"`
+	ApplicationCredentialSecret string `yaml:"application_credential_secret"`
+	Username                    string `yaml:"username"`
+	Password                    string `yaml:"password"`
+	ProjectID                   string `yaml:"project_id"`
+	ProjectName                 string `yaml:"project_name"`
+	UserDomainName              string `yaml:"user_domain_name"`
+	ProjectDomainName           string `yaml:"project_domain_name"`
+}
+
+// PropagateOpenStackSecrets propagates OpenStack secrets
+func PropagateOpenStackSecrets(ctx context.Context, cfg *PropagationCfg, credential *hmc.Credential) error {
+	// Fetch the OpenStackCluster resource
+	openstackCluster := &capo.OpenStackCluster{}
+	err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      cfg.ManagedCluster.Name,
+		Namespace: cfg.ManagedCluster.Namespace,
+	}, openstackCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("OpenStackCluster %s/%s not found: %w",
+				cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
+		}
+		return fmt.Errorf("unable to get OpenStackCluster %s/%s: %w",
+			cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
+	}
```

Comment:
```suggestion
	if err := cfg.Client.Get(ctx, client.ObjectKey{
		Name:      cfg.ManagedCluster.Name,
		Namespace: cfg.ManagedCluster.Namespace,
	}, openstackCluster); err != nil {
		return fmt.Errorf("unable to get OpenStackCluster %s/%s: %w",
			cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
	}
```

I am not sure we will benefit from duplicating the not found error (`.. not found: ... not found`)

Reply:
Removed.